components/applications-service/pkg/storage/postgres/db.go

```diff
@@ -101,10 +101,21 @@ type service struct {
 	SupID               int32     `db:"sup_id"`
 	Channel             string    `db:"channel"`
 	FullPkgIdent        string    `db:"package_ident"`
-	LastEventOccurredAt time.Time `db:"last_event_occurred_at"`
 	PreviousHealth      string    `db:"previous_health"`
+	LastEventOccurredAt time.Time `db:"last_event_occurred_at"`
+	HealthUpdatedAt     time.Time `db:"health_updated_at"`
 	CreatedAt           time.Time `db:"-"`
 	UpdatedAt           time.Time `db:"-"`
+
+	// (internal) use it to know if the service needs an update or not
+	needUpdate bool `db:"-"`
+}
+
+// implement the PreInsert hook for the service struct/table
+// TODO @afiune should we do this instead of SQL functions/triggers
```

Comment:
💭 I don't think so. This is a gorp-only thing; SQL triggers are independent of the library used to talk to the database (they're even independent of the programming language used). I think that's preferrable. Related: Do you think gorp is worth it? When we've been using that in auth land, we ended up noticing that straightforward usage of `database/sql` (with `github.com/lib/pq` of course) is not really more cumbersome, and exposes more PG features, most notably transactions.

Reply:
I just say it because to implement a timestamp update here is one line and in postgres, you have to implement the trigger and function: https://github.com/chef/automate/blob/master/components/applications-service/pkg/storage/postgres/schema/sql/04_timestamps.up.sql#L17-L43

Reply:
I'm aware of your triggers 😉 -- I'd say the disadvantage of a few extra lines is outweighted by not being tied to a specific library 🤔

Reply:
Agree! It is either or, because with the triggers and functions you have to be aware of the state of the database and maintain them. Which could be challenging in the future with X number of functions and triggers and enums, etc. etc. 😉

I love your feedback btw, this is what I was hoping for in the review process! 💯

Reply:
Is there an advantage to using gorp? With a trigger it seems you could do it all within a transaction which might be better for error handling?

Reply:
@kmacgugan I have chatted with @srenatus about the use of `gorp` and how the auth team have used it as well as the reasons why they are not using it anymore, the TL DR is that they prefer the straightforward `SELECT ...` calls and not using the package and its features, they didn't find any blocker, it was more a preference/choice.

An important point about this move is that they don't have a place they could check in the code for the state of functions, triggers, etc. in PG and they have to start the service to run all the migration scripts and check directly in the database how the definitions look, which is why I prefer having the code at hand to verify those details without the need of starting the service and database.

So, it is up to us to use it or not. I personally like the idea of thing simple things like these line to update a timestamp but when things get complicated, I would definitely want to code it as PG functions/triggers. So a mix sounds like the way to go but I would like to hear your thoughts.

Reply:
(Here's all the details 😄)

> So, we ended up not using many of gorp's features, so it didn't seem worth the dependency -- a few straightforward `SELECT ...` calls can easily be handled using the stdlib `database/sql` methods.
>
> I think at one point or another, I was annoyed by gorp for these reasons -- which may or may not have been resolved; or actually been my fault at using the library properly, or reading the docs:
>
> a) transactions -- it didn't seems straightforward to do that with gorp
> b) context handling -- I believe this was fixed by now -- any DB call should use the caller's context, we want db actions to be aborted if the client doesn't care anymore.

> It's a bit cumbersome to maintain PG functions -- migrations have to `CREATE OR REPLACE` them again and again.
>
> On the positive side, all it takes to figure out what's going on is a `chef-automate psql chef_authz_service` session, and you can query the same functions there that the go code queries. You don't have to copy-paste long queries from Go into the psql REPL. This is a weak argument, but still :sweat_smile:
>
> As a general thing, the more happens in the database, the less do we have to worry about concurrency and stuff; like, what happens if the same grpc method is executed two times concurrently.

Tl;dr: If you know the SQL, do the SQL -- we don't need fancy wrapping 😜

Reply:
@srenatus well that tl;dr is confusing since it tells me to not use it for this change and from our chat, I asked if there are blockers and/or what was your recommendation and you said:
>No, not at all. If you like working with gorp, do it. :wink: (edited) 

I think I'd be better off removing this and going SQL all the way.

Reply:
😃 sorry I wasn't explicit there -- whatever you think is best here is fine by me, I'm not going to block this! 😄 I stand by that other comment, _if you like gorp, use it_ -- what else!

The "tl;dr" was just summing up my personal preference, I guess. 😉

Reply:
Thanks for exploring this topic in depth! I think it is worth it to explore the trade offs before pulling in a new dep vs using SQL-fu.

Reply:
[pedantry] It's not really _new_ -- it's been in this code (I think), and other services in a2, for a loooong time. 😄

Reply:
Yeah, it is definitely not new. 💯